src/module/ct/event.rs

```diff
@@ -0,0 +1,176 @@
+use std::fmt;
+
+use crate::{core::events::*, event_section};
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize)]
+pub(crate) enum ZoneDir {
+    Original,
+    Reply,
+    Default,
+    #[default]
+    None,
+}
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize)]
+pub(crate) struct CtTcp {
+    /// TCP source port
+    pub(crate) sport: u16,
+    /// TCP destination port
+    pub(crate) dport: u16,
+}
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize)]
+pub(crate) struct CtUdp {
+    /// UDP source port
+    pub(crate) sport: u16,
+    /// UDP destination port
+    pub(crate) dport: u16,
+}
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize)]
+pub(crate) struct CtIcmp {
+    /// ICMP code
+    pub(crate) code: u8,
+    /// ICMP type
+    pub(crate) r#type: u8,
+    /// ICMP ID
+    pub(crate) id: u16,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub(crate) enum CtProto {
+    Tcp(CtTcp),
+    Udp(CtUdp),
+    Icmp(CtIcmp),
+}
+impl Default for CtProto {
+    fn default() -> Self {
+        CtProto::Tcp(CtTcp::default())
+    }
+}
+
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub(crate) enum CtIpVersion {
```

Comment:
I'm not sure how to build a type that would serve for both this and `skb`.

At json level, here we have
```json
"orig": {
   "src": "1.1.1.1",
    "dst": "2.2.2.2",
    "version": "v4",
}
```

In `skb` we have:
```json
"ip": {
      "daddr": "192.168.1.131",
      "saddr": "142.250.185.110",
      "v4": {
        "flags": 0,
        "id": 0,
      }
    }
```
Here we only need the version as value for key "version",  and in `skb` the version is the key for an object.
The obvious:
```
enum IPVersion <V4, V6> {
   V4(V4),
   V6(V6),
}
```
would force would force us to have:
```
type SimpleIPVersion = IPVersion<(), ()>
struct MyStruct {
    saddr: String,
    daddr: String,
    version: SimpleIPVersion,
}
```
but this would serialize to:
```json
"orig": {
   "src": "1.1.1.1",
    "dst": "2.2.2.2",
    "version": {"v4": null},
}
```
Which is not good.

So we need to implement a custom [de]serialization functions and distinguish between `IPVersion<(), ()>` and the rest. Therefore, why not just two objects `enum IPVersion {V4,V6}` and `enum IPVersionOpt<V4, V6> {V4(V4), V6(V6)}`...

Reply:
Ok. I  gave it a try: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021

WDYT?

Reply:
Also, should we do that on this PR?

Reply:
https://github.com/retis-org/retis/issues/253